src/disco/bank/fd_bank_abi.c

```diff
@@ -0,0 +1,416 @@
+    out_txn->message.legacy.message.owned.header.num_readonly_signed_accounts = txn->readonly_signed_cnt;
+    out_txn->message.legacy.message.owned.header.num_readonly_unsigned_accounts = txn->readonly_unsigned_cnt;
+    return FD_BANK_ABI_TXN_INIT_SUCCESS;
+  } else if( FD_LIKELY( txn->transaction_version==FD_TXN_V0 ) ){
+    int result = fd_ext_bank_sanitized_txn_load_addresess( bank, (void*)out_txn->message.v0.message.owned.address_table_lookups, txn->addr_table_lookup_cnt, out_sidecar );
+    if( FD_UNLIKELY( result!=FD_BANK_ABI_TXN_INIT_SUCCESS ) ) return result;
+
+    out_txn->message.v0.loaded_addresses.owned.writable_cnt = txn->addr_table_adtl_writable_cnt;
+    out_txn->message.v0.loaded_addresses.owned.writable_cap = txn->addr_table_adtl_writable_cnt;
+    out_txn->message.v0.loaded_addresses.owned.writable = (sanitized_txn_abi_pubkey_t*)out_sidecar;
+    out_sidecar += txn->addr_table_adtl_writable_cnt*32UL;
+
+    out_txn->message.v0.loaded_addresses.owned.readable_cnt = (ulong)txn->addr_table_adtl_cnt - (ulong)txn->addr_table_adtl_writable_cnt;
+    out_txn->message.v0.loaded_addresses.owned.readable_cap = (ulong)txn->addr_table_adtl_cnt - (ulong)txn->addr_table_adtl_writable_cnt;
+    out_txn->message.v0.loaded_addresses.owned.readable = (sanitized_txn_abi_pubkey_t*)out_sidecar;
+    out_sidecar += ((ulong)txn->addr_table_adtl_cnt - (ulong)txn->addr_table_adtl_writable_cnt)*32UL;
+
+    out_txn->message.v0.is_writable_account_cache_cnt = txn->acct_addr_cnt;
+    out_txn->message.v0.is_writable_account_cache_cap = txn->acct_addr_cnt;
+    out_txn->message.v0.is_writable_account_cache = out_sidecar;
+    out_sidecar += txn->acct_addr_cnt + txn->addr_table_adtl_cnt;
+
+    int _is_upgradeable_loader_present = is_upgradeable_loader_present_v0( txn, payload, (void*)out_txn->message.v0.loaded_addresses.owned.writable );
+    for( ushort i=0; i<txn->acct_addr_cnt; i++ ) {
+      int is_writable = !!fd_txn_is_writable( txn, i ) &&
+                        !is_builtin_key_or_sysvar(  payload + txn->acct_addr_off + i*32UL ) &&
+                        (!is_key_called_as_program( txn, i ) || _is_upgradeable_loader_present);
+      out_txn->message.v0.is_writable_account_cache[ i ] = !!is_writable;
+    }
+    for( ushort i=0; i<txn->addr_table_adtl_writable_cnt; i++ ) {
+      int is_writable = !is_builtin_key_or_sysvar( (uchar*)(out_txn->message.v0.loaded_addresses.owned.writable + i) ) &&
+                        (!is_key_called_as_program( txn, (ushort)(txn->acct_addr_cnt+i) ) || _is_upgradeable_loader_present);
+      out_txn->message.v0.is_writable_account_cache[ txn->acct_addr_cnt+i ] = !!is_writable;
+    }
+    for( ushort i=0; i<txn->addr_table_adtl_cnt-txn->addr_table_adtl_writable_cnt; i++ ) {
+      out_txn->message.v0.is_writable_account_cache[ txn->acct_addr_cnt+txn->addr_table_adtl_writable_cnt+i ] = 0;
+    }
+
+    out_txn->message.v0.message.owned.account_keys_cnt = txn->acct_addr_cnt;
```

Comment:
This code would get a lot cleaner if you declared a pointer to `owned` as a local variable. That would require giving the type a name though.